SU2_CFD/src/numerics/NEMO/convection/ausmplusm.cpp

```diff
@@ -0,0 +1,243 @@
+  }
+
+  sq_veli = 0.0; sq_velj = 0.0;
+  for (iDim = 0; iDim < nDim; iDim++) {
+    u_i[iDim]  = V_i[VEL_INDEX+iDim];
+    u_j[iDim]  = V_j[VEL_INDEX+iDim];
+    sq_veli   += u_i[iDim]*u_i[iDim];
+    sq_velj   += u_j[iDim]*u_j[iDim];
+  }
+
+  P_i   = V_i[P_INDEX];   P_j   = V_j[P_INDEX];
+  h_i   = V_i[H_INDEX];   h_j   = V_j[H_INDEX];
+  a_i   = V_i[A_INDEX];   a_j   = V_j[A_INDEX];
+  rho_i = V_i[RHO_INDEX]; rho_j = V_j[RHO_INDEX];
+
+  rhoCvtr_i = V_i[RHOCVTR_INDEX]; rhoCvtr_j = V_j[RHOCVTR_INDEX];
+  rhoCvve_i = V_i[RHOCVVE_INDEX]; rhoCvve_j = V_j[RHOCVVE_INDEX];
+
+  e_ve_i = 0.0; e_ve_j = 0.0;
+  for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
+    e_ve_i += (rhos_i[iSpecies]*eve_i[iSpecies])/rho_i;
+    e_ve_j += (rhos_j[iSpecies]*eve_j[iSpecies])/rho_j;
+  }
+
+  /*--- Projected velocities ---*/
+  ProjVel_i = 0.0; ProjVel_j = 0.0;
+  for (iDim = 0; iDim < nDim; iDim++) {
+    ProjVel_i += u_i[iDim]*UnitNormal[iDim];
+    ProjVel_j += u_j[iDim]*UnitNormal[iDim];
+  }
+
+  sqVi = 0.0;   sqVj = 0.0;
+  for (iDim = 0; iDim < nDim; iDim++) {
+    sqVi += (u_i[iDim]-ProjVel_i*UnitNormal[iDim]) *
+            (u_i[iDim]-ProjVel_i*UnitNormal[iDim]);
+    sqVj += (u_j[iDim]-ProjVel_j*UnitNormal[iDim]) *
+            (u_j[iDim]-ProjVel_j*UnitNormal[iDim]);
+  }
+
+  /*--- Calculate interface numerical gammas and speed of sound ---*/
+  //Hnorm = 0.5*(h_i-0.5*sqVi + h_j-0.5*sqVj); --> This formula worked until V. 7.3.0
+  Hnorm = 0.5*(h_i + h_j);
+  gtl_i = Gamma_i;
+  gtl_j = Gamma_j;
+  gam   = 0.5*(gtl_i+gtl_j);
+  atl = sqrt(2.0*Hnorm*(gam-1.0)/(gam+1.0));
+
+  // ATL has to account sound speed for reactive gases
+
+ /* if (fabs(rho_i-rho_j)/(0.5*(rho_i+rho_j)) < 1E-3)   // Why this If condition ?
+    atl = sqrt(2.0*Hnorm*(gam-1.0)/(gam+1.0));
+  else {
+    //Is the bottom equation correct ??
+    atl = sqrt(2.0*Hnorm * (((gtl_i-1.0)/(gtl_i*rho_i) - (gtl_j-1.0)/(gtl_j*rho_j))/
+                            ((gtl_j+1.0)/(gtl_j*rho_i) - (gtl_i+1.0)/(gtl_i*rho_j))));
+
+
+    // Or should it be
+    atl = sqrt(2.0*Hnorm * (((gtl_i-1.0)/(gtl_i)*rho_i - (gtl_j-1.0)/(gtl_j)*rho_j)/
+                            ((gtl_j+1.0)/(gtl_j)*rho_i - (gtl_i+1.0)/(gtl_i)*rho_j)));
+  } */
+
+  if (0.5*(ProjVel_i+ProjVel_j) >= 0.0) aij = atl*atl/max(fabs(ProjVel_i),atl);
+  else                                  aij = atl*atl/max(fabs(ProjVel_j),atl);
+
+  aF=aij;
+
+  mL  = ProjVel_i/aF;
+  mR  = ProjVel_j/aF;
+
+  rhoF = 0.5*(rho_i+rho_j);
+  MFsq = 0.5*(sq_veli+sq_velj)/(aF*aF);
+
+  param1 = max(MFsq, Minf*Minf);
+  Mrefsq = (min(1.0, param1));
+  fa = 2.0*sqrt(Mrefsq)-Mrefsq;
+
+  alpha = 3.0/16.0*(-4.0+5.0*fa*fa);
+  beta = 1.0/8.0;
+
+  /*--- Pressure diffusion term ---*/
+
+  h_k=1.0;
+
+  for (auto Point_aux : nemo_geometry->nodes->GetPoints(jPoint)) {
+    P_k = nemo_solution->GetPrimitive(Point_aux,P_INDEX)/nemo_solution->GetPrimitive(jPoint,P_INDEX);
+    h_k = min(h_k,P_k);
+    h_k = min(h_k,1/P_k);
+  }
```

Comment:
This is not correct for MPI, when either iPoint or jPoint is an halo, you do not have all their neighbors to compute the sensor correctly. You need to compute the sensor over nPointDomain and then communicate, as we do for other schemes.
This way you also avoid more virtual functions in CNumerics since there is already a "SetSensor" somewhere.

Reply:
Used the sensors to pass the Pressure term to the convective scheme